src/lib.rs

```diff
@@ -24,6 +35,10 @@ pub struct Report {
     pub footer: Option<ReportFooter>,
 }
 
+pub trait Diagnosable {
+    fn diagnose(&self, source_map: &SourceMap) -> Diagnostics;
+}
+
```

Comment:
Hm, now thinking about it, this means that `Diagnostic` would be exposed as part of the public API.

what if we do this instead:

```suggestion
pub trait Reportable {
    fn report(&self) -> Vec<Report>;
}
```

this would allow mean we don't have to expose codespan to the public API.